inferno-ml-remote/src/Inferno/ML/Remote/Types.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Inferno.ML.Remote.Types
+  ( InfernoMlRemoteAPI,
+    Options (Options),
+    Script (Script),
+    EvalResult (EvalResult),
+    InfernoMlRemoteM,
+    InfernoMlRemoteEnv (InfernoMlRemoteEnv),
+    ModelCacheOption (..),
+    ModelCache (ModelCache),
+    SomeInfernoError (..),
+    InfernoMlRemoteError (..),
+    parseOptions,
+    mkOptions,
+  )
+where
+
+import Control.Applicative (Alternative ((<|>)), optional, (<**>))
+import Control.Exception (Exception (displayException), throwIO)
+import Control.Monad.Reader (ReaderT)
+import Data.Aeson
+  ( FromJSON (parseJSON),
+    ToJSON,
+    eitherDecodeFileStrict,
+    withObject,
+    (.:),
+    (.:?),
+  )
+import Data.String (IsString)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Word (Word64)
+import GHC.Generics (Generic)
+import qualified Options.Applicative as Options
+import Servant (Handler, JSON, Post, ReqBody, (:>))
+
+type InfernoMlRemoteAPI =
+  "inference" :> ReqBody '[JSON] Script :> Post '[JSON] EvalResult
```

Comment:
The input type could/should be `Expr` instead of `Text`, as I'm guessing scripts will be saved in inferno-vc, which stores scripts as the `Expr` ASTs instead of as text. That way you can also avoid parsing the scripts for every request.

Reply:
A similar thing applies to the comment above. There's no `FromJSON` instance, but if we can write a lawful, sensible instance then I'd also prefer that

Reply:
I think we do have one, don't we? `inferno-vc` stores `Expr`s as JSON. Does this not work?
https://github.com/plow-technologies/inferno/blob/2ca4041cb4c9dfe611b051edde59430ea6ad1942/inferno-types/src/Inferno/Types/Syntax.hs#L719

Reply:
Oh I must have missed that. We'd have to change how several things currently work. I'd prefer to merge this first and then figure that out